src/libraries/Microsoft.Extensions.Options/src/Microsoft.Extensions.Options.csproj

```diff
@@ -14,6 +14,7 @@
 
   <ItemGroup>
     <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.DependencyInjection.Abstractions\src\Microsoft.Extensions.DependencyInjection.Abstractions.csproj" />
+    <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.Hosting.Abstractions\src\Microsoft.Extensions.Hosting.Abstractions.csproj" />
```

Comment:
Do we really want this layering? Can't I use Options without Hosting?

Reply:
Should it be in the hosting assembly?

Reply:
Why do we need a hosted service at all? That’s the part I’m not understanding

Reply:
Yes, the goal of this feature is to run validation code on start. That can only be done if something resolved the options value. In that case, that thing doing the resolution is the hosted service.

Reply:
But only on "start" of a Host? Is there a way to get code executed right after the DI container is built?

I'm concerned with putting this feature into the "Options" assembly, but it doesn't work if I just use "Options" and I don't use Hosting.

Reply:
This only works for the hosting model, there's no other way to "run code at startup". Maybe this should be in hosting.

Reply:
DI doesn't have a hook that I can run some code after the container is built?

Reply:
No, it doesn't have that. But even if it did, we can't have options tied to the specific our container implementation. The other feature request we have gotten though is to allow pre-warming singletons. That's something that could be in the container. But it wouldn't help here, you'd need to add another singleton that resolved the options type in the constructor (which I guess could work?)

Reply:
#### Regarding assembly:
Based on the above discussion it makes sense to move the diff onto the hosting assembly. 

#### Regarding namespace:
The namespace for ValidateOnStart was approved under Microsoft.Extensions.DependencyInjection in the .NET API review due to reasons:

1. because it is used along with configuring the IServiceCollection. In the ASP.NET Core templates' Startup.cs classes we have a using for Microsoft.Extensions.DependencyInjection but not Microsoft.Extensions.Options
2. keeping consistent with the most common APIs related to OptionsBuilder:

https://github.com/dotnet/runtime/blob/1d9e50cb4735df46d3de0cee5791e97295eaf588/src/libraries/Microsoft.Extensions.Options.DataAnnotations/ref/Microsoft.Extensions.Options.DataAnnotations.cs#L11

https://github.com/dotnet/runtime/blob/1d9e50cb4735df46d3de0cee5791e97295eaf588/src/libraries/Microsoft.Extensions.Options/ref/Microsoft.Extensions.Options.cs#L12-L13

https://github.com/dotnet/runtime/blob/1d9e50cb4735df46d3de0cee5791e97295eaf588/src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/ref/Microsoft.Extensions.Options.ConfigurationExtensions.cs#L11-L13

cc: @halter73